components/platform/platform.c

```diff
@@ -65,60 +83,125 @@ uart_status_t uart_status[NUM_UART];
+{
   uart_event_post_t *post = (uart_event_post_t *)param;
   unsigned id = post->id;
   uart_status_t *us = &uart_status[id];
+  UNADJUST_ID(id);
   xSemaphoreGive(sem);
-  if(post->type == PLATFORM_UART_EVENT_DATA) {
-    size_t i = 0;
-    while (i < post->size)
-    {
-      if (id == CONFIG_ESP_CONSOLE_UART_NUM && run_input) {
-        unsigned used = feed_lua_input(post->data + i, post->size - i);
-        i += used;
-      }
-      else {
-        char ch = post->data[i];
-        us->line_buffer[us->line_position] = ch;
-        us->line_position++;
-
-        uint16_t need_len = us->need_len;
-        int16_t end_char = us->end_char;
-        size_t max_wanted =
-          (end_char >= 0 && need_len == 0) ? LUA_MAXINPUT : need_len;
-        bool at_end = (us->line_position >= max_wanted);
-        bool end_char_found =
-          (end_char >= 0 && (uint8_t)ch == (uint8_t)end_char);
-        if (at_end || end_char_found) {
-          uart_on_data_cb(id, us->line_buffer, us->line_position);
-          us->line_position = 0;
-        }
-        ++i;
-      }
-    }
+  if (post->type == PLATFORM_UART_EVENT_DATA)
+  {
+    handle_uart_data(id, post, us);
     free(post->data);
-  } else {
+  }
+  else
+  {
     const char *err;
-    switch(post->type) {
-      case PLATFORM_UART_EVENT_OOM:
-        err = "out_of_memory";
-        break;
-      case PLATFORM_UART_EVENT_BREAK:
-        err = "break";
-        break;
-      case PLATFORM_UART_EVENT_RX:
-      default:
-        err = "rx_error";
+    switch (post->type)
+    {
+    case PLATFORM_UART_EVENT_OOM:
+      err = "out_of_memory";
+      break;
+    case PLATFORM_UART_EVENT_BREAK:
+      err = "break";
+      break;
+    case PLATFORM_UART_EVENT_RX:
+    default:
+      err = "rx_error";
     }
     uart_on_error_cb(id, err, strlen(err));
   }
   free(post);
 }
 
+#if CONFIG_ESP_CONSOLE_USB_SERIAL_JTAG
+void usbcdc_event_task(task_param_t param, task_prio_t prio)
+{
+  uart_event_post_t *post = (uart_event_post_t *)param;
+  xSemaphoreGive(usbcdc_sem);
+  handle_uart_data(0, post, &usbcdc_status);
+  free(post);
+}
+
+static void task_usbcdc(void *pvParameters) {
+  // 4 chosen as a number smaller than the number of nodemcu task slots
+  // available, to make it unlikely we encounter a task_post failing
+  if (usbcdc_sem == NULL)
+    usbcdc_sem = xSemaphoreCreateCounting(4, 4);
+
+  uart_event_post_t *post = NULL;
+
+  for (;;) {
+    if (post == NULL) {
+      post = (uart_event_post_t *)malloc(sizeof(uart_event_post_t) + 64 * sizeof(char));
+      if (post == NULL) {
+        ESP_LOGE(UART_TAG, "Can not alloc memory in task_usbcdc()");
+        // reboot here?
+        continue;
+      }
+      post->data = (void *)(post + 1);
+      post->type = PLATFORM_UART_EVENT_DATA;
+    }
+
+    int len = usb_serial_jtag_read_bytes(post->data, 64, portMAX_DELAY);
+
+    if (len > 0) {
+      for (int i = 0; i < len; i++) {
+        if (post->data[i] == '\r') {
+          post->data[i] = '\n';
```

Comment:
Is there any chance of having the CDC serial active but not be the console? If so, we probably should not be doing this translation here. Also, won't this result in a double-prompt if someone does send a \r\n? As in
```
> =node.heap()\r\n
123456
>
>
```

Reply:
I removed this translation as it wasn't needed -- I had it there while I was trying to make stuff work!